examples/fine-tuned-RAG/ModelFinetune.ipynb

```diff
@@ -0,0 +1,1186 @@
+     ]
+    }
+   ],
+   "source": [
+    "# Use progress_apply with tqdm for progress bar\n",
+    "df[\"generated_answer\"] = df.progress_apply(answer_question, axis=1)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df.to_json(\"local_cache/100_val.json\", orient=\"records\", lines=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Fine-Tuning the OpenAI Model\n",
+    "\n",
+    "For the complete fine-tuning process, please refer to the [OpenAI Fine-Tuning Docs](https://platform.openai.com/docs/guides/fine-tuning/use-a-fine-tuned-model).\n",
+    "\n",
+    "### Prepare the Fine-Tuning Data\n",
+    "\n",
+    "We need to prepare the data for fine-tuning. We'll use a few samples from train split of same dataset as before, but we'll add the answer to the context. This will help the model learn to retrieve the answer from the context."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "100%|██████████| 100/100 [00:00<00:00, 68067.25it/s]\n"
+     ]
+    }
+   ],
+   "source": [
+    "def dataframe_to_jsonl(df):\n",
+    "    def create_jsonl_entry(row):\n",
+    "        answer = row[\"answers\"][0] if row[\"answers\"] else \"I don't know\"\n",
+    "        messages = [\n",
+    "            {\"role\": \"system\", \"content\": \"You are a helpful assistant.\"},\n",
+    "            {\n",
+    "                \"role\": \"user\",\n",
+    "                \"content\": f\"\"\"Answer the following Question based on the Context only. Only answer from the Context. If you don't know the answer, say 'I don't know'.\n",
+    "            Question: {row.question}\\n\\n\n",
+    "            Context: {row.context}\\n\\n\n",
+    "            Answer:\\n\"\"\",\n",
+    "            },\n",
+    "            {\"role\": \"assistant\", \"content\": answer},\n",
+    "        ]\n",
+    "        return json.dumps({\"messages\": messages})\n",
+    "\n",
+    "    jsonl_output = df.progress_apply(create_jsonl_entry, axis=1)\n",
+    "    return \"\\n\".join(jsonl_output)\n",
+    "\n",
+    "train_sample = get_diverse_sample(train_df, sample_size=100, random_state=42)\n",
+    "\n",
+    "with open(\"local_cache/100_train.jsonl\", \"w\") as f:\n",
+    "    f.write(dataframe_to_jsonl(train_sample))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### [Optional] Verify the Fine-Tuning Data\n",
+    "\n",
+    "You can see this [cookbook](https://github.com/openai/openai-cookbook/blob/main/examples/Chat_finetuning_data_prep.ipynb) for more details on how to prepare the data for fine-tuning."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Fine-Tune OpenAI Model\n",
```

Comment:
This header seems redundant with the one a couple cells above - can you merge these into one "fine-tune an OpenAI Chat Model" section